### ISSUE TITLE
Drizzle and mapDispatchToProps not working

### ISSUE THREAD

**Issue by sarbogast** [ConsenSys-archive/drizzle-legacy#112]
I'm still struggling with Drizzle, trying to dispatch my own action. Based on the Drizzle box, I added the following to HomeContainer:

```
const mapDispatchToProps = dispatch => {
    return {
        onGetEthRate: (fiatSymbol) => dispatch(actions.getEthRate(fiatSymbol))
    };
};
const HomeContainer = drizzleConnect(Home, mapStateToProps, mapDispatchToProps);
```
The getEthRate action creator looks like this:

```
export const getEthRate = (fiatSymbol) => {
    console.log("Getting ETH rate for " + fiatSymbol);
    return {
        type: actionTypes.GET_ETH_RATE,
        fiatSymbol: fiatSymbol
    };
};
```
And then I registered a saga to process that action in rootSaga.js:

```
export default function* root() {
    yield all(
        drizzleSagas.map(saga => fork(saga)),
        takeEvery(actionTypes.GET_ETH_RATE, getEthRateSaga)
    );
}
```
But the following saga is never triggered, as I never see the log in the console:

```
export function* getEthRateSaga(action) {
    console.log("Fetching ETH rate for " + action.fiatSymbol);
    try {
        const response = yield http.get('/price?fsym=ETH&tsyms=' + action.fiatSymbol);
        const rate = response.data[action.fiatSymbol];
        yield put(actions.updateEthRate(action.fiatSymbol, rate));
    } catch(error) {
        console.error(error);
    }
}
```
And based on the Redux dev console, the GET_ETH_RATE action is not even dispatched, so I'm wondering if mapDispatchToProps works with drizzleConnect.

**Comment by reime005** [comment]
You have to make sure that you initialized everything, including reducers, sagas and then initialize Drizzle. I made an example here, which may help you: [https://github.com/reime005/react-redux-saga-drizzle](https://github.com/reime005/react-redux-saga-drizzle).

I think you don't really need drizzleConnect

**Comment by sarbogast** [comment]
I think I have double checked all my initialization several times, and obviously I'm missing something. Classical case of "it's right in front of my eyes".
Here is the content of my index.js:

```
import React from 'react';
import ReactDOM from 'react-dom';
import { Router, Route, IndexRoute, browserHistory } from 'react-router'
import { syncHistoryWithStore } from 'react-router-redux'
import { DrizzleProvider } from 'drizzle-react'

// Layouts
import App from './App'
import HomeContainer from './layouts/home/HomeContainer'
import { LoadingContainer } from 'drizzle-react-components'

import store from './store'
import drizzleOptions from './drizzleOptions'

// Initialize react-router-redux.
const history = syncHistoryWithStore(browserHistory, store)

ReactDOM.render((
    <DrizzleProvider options={drizzleOptions} store={store}>
      <LoadingContainer>
        <Router history={history}>
          <Route path="/" component={App}>
            <IndexRoute component={HomeContainer} />
          </Route>
        </Router>
      </LoadingContainer>
    </DrizzleProvider>
  ),
  document.getElementById('root')
);

```

Didn't change anything compared to the drizzle box code, since the real things happen in store.js:

```
import { browserHistory } from 'react-router'
import { createStore, applyMiddleware, compose } from 'redux'
import thunkMiddleware from 'redux-thunk'
import { routerMiddleware } from 'react-router-redux'
import reducer from './reducer'
import rootSaga from './rootSaga'
import createSagaMiddleware from 'redux-saga'
import { generateContractsInitialState } from 'drizzle'
import drizzleOptions from './drizzleOptions'

// Redux DevTools
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const routingMiddleware = routerMiddleware(browserHistory);
const sagaMiddleware = createSagaMiddleware();

const initialState = {
  contracts: generateContractsInitialState(drizzleOptions)
};

const store = createStore(
  reducer,
  initialState,
  composeEnhancers(
    applyMiddleware(
      thunkMiddleware,
      routingMiddleware,
      sagaMiddleware
    )
  )
);

sagaMiddleware.run(rootSaga);

export default store;
```

And here is the content of reducer.js, where I add my own reducer:
```
import {combineReducers} from 'redux'
import {routerReducer} from 'react-router-redux'
import {drizzleReducers} from 'drizzle'
import ratesReducer from './store/reducers/rates';

const reducer = combineReducers({
    routing: routerReducer,
    rates: ratesReducer,
    ...drizzleReducers
});

export default reducer;
```

And the content of rootSaga.js, where I add a saga to process the GET_ETH_RATE action:
```
import {all, fork, takeEvery} from 'redux-saga/effects'
import {drizzleSagas} from 'drizzle'
import {getEthRateSaga} from "./store/sagas/sagas";
import * as actionTypes from "./store/actions/actionTypes";

export default function* root() {
    yield all(
        drizzleSagas.map(saga => fork(saga)),
        takeEvery(actionTypes.GET_ETH_RATE, getEthRateSaga)
    );
}
```

What seems particularly impossible to me is that my getEthRate() function gets called, I can see the console log in it, but the action is returned seems not to be dispatched at all.

**Comment by reime005** [comment]
You could try to use the react-redux provider instead (as I did) and/or then call the Drizzle constructor, which should dispatch the initialization actions on your store. I did it like this so I don't need `drizzleConnect` or `DrizzleProvider`. This might be your problem somehow

**Comment by sarbogast** [comment]
OK, so I figured it out. I was misled by Redux DevTools that made me think that the action was not dispatched because when I filtered on the action's name, I didn't see anything. But apparently, the action was dispatched, but not picked up by Saga because there was a mistake in my initialization there:

```
export default function* root() {
    yield all(
        drizzleSagas.map(saga => fork(saga)),
        takeEvery(actionTypes.GET_ETH_RATE, getEthRateSaga)
    );
}
```

The issue is that all() expects an array, and I pass it an array and an isolated object. Here is the right syntax to fix that and append my takeEvery to the result of drizzleSagas.map():

```
export default function* root() {
    yield all([
        ...drizzleSagas.map(saga => fork(saga)),
        takeEvery(actionTypes.GET_ETH_RATE, getEthRateSaga)
    ]);
}
```

With that, everything works fine.
